spec/lib/miq_expression/query_helper_spec.rb

```diff
@@ -0,0 +1,142 @@
+RSpec.describe MiqExpression::QueryHelper do
+  describe ".remove_equality" do
+    let(:vm_host_id) { Vm.arel_table[:host_id] }
+    let(:host_id) { Host.arel_table[:id] }
+    let(:host_type) { Host.arel_table[:type] }
+    let(:bind_param) { Arel::Nodes::BindParam.new }
+
+    # query:  nil
+    # result: nil
+    it "handles an empty query" do
+      query = nil
+      result = described_class.remove_equality(query, vm_host_id.eq(host_id))
+      expect(result).to be_nil
+    end
+
+    # query:  "vms"."host_id" = "hosts"."id"
+    # result: nil
+    it "removes a matching query" do
+      query = vm_host_id.eq(host_id)
+      result = described_class.remove_equality(query, vm_host_id.eq(host_id))
+      expect(result).to be_nil
+    end
+
+    # query:  "hosts"."id" = "vms"."host_id"
+    # result: nil
+    it "removes a matching query (opposite order)" do
+      query = host_id.eq(vm_host_id)
+      result = described_class.remove_equality(query, vm_host_id.eq(host_id))
+      expect(result).to be_nil
+    end
+
+    # query:  "hosts"."type" = 'abc'
+    # result: "hosts"."type" = 'abc'
+    it "keeps a non matching query" do
+      query = host_type.eq('abc')
+      result = described_class.remove_equality(query, vm_host_id.eq(host_id), [])
+      expect(result).to eq(query)
+    end
+
+    # query:  "hosts"."type" = ?
+    # result: "hosts"."type" = 5
+    it "keeps a non matching bind query" do
+      query = host_type.eq(bind_param)
+      expected = host_type.eq(Arel::Nodes::Casted.new(5, host_type))
+      result = described_class.remove_equality(query, vm_host_id.eq(host_id), bind_values(5))
+      expect(result).to eq(expected)
+    end
+
+    # query:  "hosts"."type" = 'abc'
+    # result: "hosts"."type" = 'abc'
+    # This doesn't make sense, but we saw this example
+    it "removes a matching query in an AND (single node)" do
+      query = arel_and(vm_host_id.eq(host_id))
+      result = described_class.remove_equality(query, vm_host_id.eq(host_id))
+      expect(result).to be_nil
+    end
+
+    # query:  <invalid>
```

Comment:
I feel like `<invalid>` is serving much of a purpose as to what `<invalid>` is supposed to be, or where it might happen.  Again, similar comment to the "but we saw this example" one.